weave/flow/eval_imperative.py

```diff
@@ -0,0 +1,246 @@
+            # return score
+            # TODO: can't use score here because it will cause version mismatc
+            return cast(float, current_score.get())
+
+        scorer_instance.__dict__["score"] = MethodType(score_method, scorer_instance)
+
+        # attach the score feedback to the predict call
+        with call_context.set_call_stack(
+            [self.evaluate_call, self.predict_and_score_call]
+        ):
+            with set_current_score(score):
+                await self.predict_and_score_call.apply_scorer(scorer_instance)
+
+
+class ImperativeEvaluationLogger(BaseModel):
+    """This class provides an imperative interface for logging evaluations.
+
+    An evaluation is started automatically when the first prediction is logged
+    using the `log_prediction` method, and finished when the `log_summary` method
+    is called.
+
+    Each time you log a prediction, you will get back an `ImperativePredictionLogger`
+    object.  You can use this object to log scores and metadata for that specific
+    prediction (see that class for more details).
+
+    Example:
+        ```python
+        ev = ImperativeEvaluationLogger()
+        pred = ev.log_prediction(inputs, output)
+        await pred.log_score(scorer_name, score)
+        ev.log_summary(summary)
+        ```
+    """
+
+    _eval_started: bool = PrivateAttr(False)
+    _logged_summary: bool = PrivateAttr(False)
+    _evaluate_call: Call | None = PrivateAttr(None)
+    _pseudo_model: Model = PrivateAttr(default_factory=lambda: Model())
+    _pseudo_evaluation: Evaluation = PrivateAttr(
+        default_factory=lambda: Evaluation(
+            dataset=weave.Dataset(rows=weave.Table([{"": ""}])),
+            scorers=[],
+        )
+    )
+    _starting_stack: list[Call] = PrivateAttr(default_factory=list)
+
+    def log_prediction(self, inputs: dict, output: Any) -> ImperativeScoreLogger:
+        if not self._eval_started:
+            self._eval_started = True
+
+            # The following section is a "hacky" way to create Model and Evaluation
+            # objects that "look right" to our object saving system.
+
+            # --- Setup the model object ---
+            @weave.op
+            def predict(self: Model, inputs: dict) -> Any:
+                # Get the output from the context variable
+                return current_output.get()
+
+            self._pseudo_model.__dict__["predict"] = MethodType(
+                predict, self._pseudo_model
+            )
+
+            # --- Setup the evaluation object ---
+            @weave.op(name="Evaluation.evaluate")
+            def evaluate(self: Evaluation, model: Model) -> None: ...
+
+            @weave.op
+            def predict_and_score(self: Evaluation, model: Model, inputs: dict) -> dict:
+                model_output = model.get_infer_method()(inputs)
+                return {
+                    "model_output": model_output,
+                    "scores": NOT_DEFINED,
```

Comment:
The `NOT_DEFINED` value: a) looks funky in the UI; b: deviates from the standard type. I would prefer either None, or empty dict/0 respectively.